### ISSUE TITLE
Feature Request: Extract DOI numbers and download academic paper PDFs automatically

### ISSUE THREAD

**Issue by pirate** [ArchiveBox/ArchiveBox#720]
New extractor idea: SCIHUB

e.g. take this academic paper for example: https://www.cell.com/current-biology/fulltext/S0960-9822(19)31469-1

If a full paper PDF is available on scihub, e.g.: https://sci-hub.se/https://www.cell.com/current-biology/fulltext/S0960-9822(19)31469-1 it could be downloaded to a `./archive/<timestmap>/scihub/` output folder.

```bash
# try downloading via verbatim URL first
$ scihub.py -d https://www.cell.com/current-biology/fulltext/S0960-9822(19)31469-1'
DEBUG:Sci-Hub:Successfully downloaded file with identifier https://www.cell.com/current-biology/fulltext/S0960-9822(19)31469-1
```

We could also look for a DOI number in the page URL or page html contents e.g.: `10.1016/j.cub.2019.11.030` using a regex and try downloading that.
```bash
# otherwise try downloading via any regex-extracted bare DOI numbers on the page or in the URL
$ scihub.py -d '10.1016/j.cub.2019.11.030'
DEBUG:Sci-Hub:Successfully downloaded file with identifier 10.1016/j.cub.2019.11.030

$ ls
c28dc1242df6f931c29b9cd445a55597-.cub.2019.11.030.pdf
```

**New Dependencies:**
- https://github.com/zaytoun/scihub.py

**New Extractors:**
- `extractors/scihub.py`

**New Config Options:**
- `SAVE_SCIHUB=True`

**Comment by lbcnz** [comment]
Captchas would be a problem.

**Comment by pirate** [comment]
In my testing so far it's not a problem as long as you're not doing dozens of papers at a time.

**Comment by jezcope** [comment]
Two more avenues to explore:

1. Zotero has a big collection of translators to get full text from many different publishers and repositories https://github.com/zotero/translators
2. Unpaywall has a huge database of open access copies of articles https://unpaywall.org/

Would be great to capture full text for academic publications where its available.

**Comment by pirate** [comment]
Parsing DOIs out of PDF/HTML content: https://github.com/MicheleCotrufo/pdf2doi

Found good candidate extractor dependencies to try for all the major free scientific paper DBs:

- https://github.com/Tishacy/SciDownl
- https://github.com/viown/libgen-dl
- https://github.com/unpywall/unpywall
- https://openaccessbutton.org/?from=integration&id=
- https://github.com/bibcure/scihub2pdf

and more via...
- https://gist.github.com/taskylizard/5ba73bf97dccf159316edcf4c6520856#-academic-papers
- https://github.com/cipher387/osint_stuff_tool_collection
- https://www.reddit.com/r/FREEMEDIAHECKYEAH/wiki/tools-index

something like this might be interesting for linking together citations too https://inciteful.xyz/

**Comment by benmuth** [comment]
I have a rough script for this working (just using `scihub.py` as a module and downloading a pdf with the url->doi fallback). I had to slightly modify `scihub.py` to get it to work, but I was thinking we can just add the modified version to the `vendor` directory or something. It looks like `scihub.py` isn't available on PyPI anyway.

Should I just wait for the plugin system? Or I can create a repo with the script and modified `scihub.py` so it can be looked at and improved in the meantime. Or I can try to minimally integrate it with ArchiveBox (just get it working as a `oneshot`?) and open a PR for feedback and testing, whatever makes sense.

**Comment by pirate** [comment]
If you can post it in a gist and share the link that would be helpful @benmuth!

**Comment by benmuth** [comment]
Here's what I have so far https://gist.github.com/benmuth/b2c12cbb40ca4d8183c6f17f819e2f2d @pirate

**Comment by pirate** [comment]
@benmuth it might take me a month or more till I'm able to merge this, as I'm working on some paid ArchiveBox projects right now for a client that are taking up most of my time. Don't worry, I'm keen to merge it though, I've been wanting this feature personally for a long time.

**Comment by benmuth** [comment]
> @benmuth it might take me a month or more till I'm able to merge this, as I'm working on some paid ArchiveBox projects right now for a client that are taking up most of my time. Don't worry, I'm keen to merge it though, I've been wanting this feature personally for a long time.

No worries! In the meantime let me know if there's anything I can do to enhance it. I've kept it simple for demonstration purposes, but it seems straightforward to add some of the ideas you linked here, like the PDF parsing.

**Comment by pirate** [comment]
@benmuth after thinking about this a bit more, I think it could be fun to release your `scihub-dl` script as its own dedicated PyPI package!

Publishing a package is great resume candy and it'll make this tool usable to a much wider audience than just ArchiveBox users. Then we can just add it to `pyproject.toml` as a requirement for ArchiveBox and call it like any other command-line tool extractor.

Here are some examples as inspiration for CLI interface, PyPI packaging setup, README, etc.:
- https://pypi.org/project/arxiv-dl/
- https://pypi.org/project/scidownl/
- https://github.com/dheison0/annas-archive-api

Your tool could even call out to these and others ^ if you want to be an all-in-one paper / research downloader.

Here's how I'm imagining it could work (but not set in stone at all, just a rough idea):

- `research-dl parse [--extract=doi,pmid,issn,isbn,title,url] ./some_archived_page.html`  
  expose a parsing CLI that can take a text/html file or URL and search through it to find any DOI/PMID/ISSN/ISBN/etc. paper URLs it finds within

- `research-dl search [--doi '10.1145/3375633'] [--pmid '...'] [--title 'some title'] [--url 'https://doi.org/10.1145/3375633']`  
  expose a paper searching and metadata retrieval CLI that can search by DOI/PMID/etc./title/URL and return metadata like `title`, `authors`, `description`, `publishDate`, `downloadURLs`, etc. for each result
  
- `research-dl download [--extract=pdf,html] [--doi '10.1145/3375633']`  
  expose paper downloading CLI that can take an identifier and download the corresponding document as PDF/html/epub/etc.
  - `--provider=auto` = "by any means necessary" (try all the scihub mirrors, check arxiv-dl, try annas-archive, google scholar, etc.)
  - `--provider=arxiv.org,sci-hub.ru,...` try specific scihub mirror, annas-archive mirror, libgen server, etc. other provider
  - `--extract=pdf,html,epub,md`
  
What do you think? Does it sound like a project that you'd be excited about? If not, no worries we can go the simpler route of just integrating your script as-is.

**Comment by benmuth** [comment]
@pirate Yeah, I think that's a great idea, I'd be happy to try to work on this. I think a more comprehensive tool should definitely exist. Thanks for the overview, that's really helpful.  

As for ownership, I'm not really sure. Maybe I can start it and I can transfer ownership to ArchiveBox if it grows and the maintenance burden proves too much for whatever reason? I don't feel strongly about it either.

I'll start a repo with what I have so far soon. I think I'll go with `papers-dl` or `sci-dl` for now.

**Comment by amosip** [comment]
Is this happening?  I would be very interested in using a tool like this.
